clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1151,10 +1162,113 @@ public void assign(Collection<TopicPartition> partitions) {
     public ConsumerRecords<K, V> poll(final Duration timeout) {
-        return poll(timeout.toMillis(), true);
+        return poll(timeout.toMillis(), true, true);
     }
 
-    private ConsumerRecords<K, V> poll(final long timeoutMs, final boolean includeMetadataInTimeout) {
+    private void getStartAndEndOffsets() {
+        coordinator.ensureActiveGroup();
+        for (TopicPartition partition : this.subscriptions.assignedPartitions()) {
+            OffsetAndMetadata offsetMetadata = committed(partition, Duration.ofMillis(requestTimeoutMs), true);
+            startOffsets.put(partition, offsetMetadata == null ? new OffsetAndMetadata(0L) : offsetMetadata);
+        }
+        seekToEnd(Collections.EMPTY_LIST);
+        for (TopicPartition partition : this.subscriptions.assignedPartitions()) {
+            endOffsets.put(partition, position(partition));
+        }
+    }
+
+    private void checkRebalance() {
+        if (useParallelRebalance) {
+            if (rebalanceConsumer == null) {
+                rebalanceConsumer = new RebalanceKafkaConsumer(configs,
+                                                               null,
+                                                               null,
+                                                               new HashMap<>(),
+                                                               new HashMap<>());
+                consumerThread = new Thread(rebalanceConsumer);
+            }
+            if (coordinator.isRebalancing(false)) {
+                System.out.println("Starting thread");
+                getStartAndEndOffsets();
+                rebalanceConsumer.addNewOffsets(startOffsets, endOffsets);
+                rebalanceConsumer.setNewQueue(coordinator.getQueue());
+                if (!consumerThread.isAlive()) {
+                    System.out.println("consumerThread has started");
+                    consumerThread.start();
+                }
+            }
+        }
+    }
+
+    private void checkRebalanceAndPoll(long timeoutMs) {
+
+        final long now = time.milliseconds();
+        checkRebalance();
+        final long elapsedMs = time.milliseconds() - now;
+
+        if (consumerThread != null && consumerThread.isAlive()) {
+            rebalanceConsumer.sendRequest(Duration.ofMillis(remainingTimeAtLeastZero(timeoutMs, elapsedMs)),
+                                          RebalanceKafkaConsumer.ConsumerRequest.POLL,
+                             null,
+                                          new DefaultTaskCompletionCallback());
+        }
+    }
+
+    private RebalanceKafkaConsumer.RequestResult pollForResults(final long timeoutMs, final long now) {
+        long elapsed = time.milliseconds() - now;
+        boolean condition = remainingTimeAtLeastZero(timeoutMs, elapsed) != 0;
+        while (result == null && condition) {
+            try {
+                Thread.sleep(retryBackoffMs);
+            } catch (InterruptedException exc) { }
```

Comment:
Restore interrupt status by calling Thread.currentThread().interrupt();

Reply:
Oh, sure I will do that.